lib/reply.js

```diff
@@ -140,27 +143,36 @@ function wrapOnSendEnd (err, reply, payload) {
 }
 
 function onSendEnd (reply, payload) {
+  const res = reply.res
+
   if (payload === undefined || payload === null) {
     reply.sent = true
-    reply.res.end()
+    reply._headers['content-length'] = '0'
+    res.writeHead(res.statusCode, reply._headers)
+    // avoid ArgumentsAdaptorTrampoline from V8
+    res.end(null, null, null)
     return
   }
 
   if (typeof payload.pipe === 'function') {
-    sendStream(payload, reply.res, reply)
+    sendStream(payload, res, reply)
     return
   }
 
   if (typeof payload !== 'string' && !Buffer.isBuffer(payload)) {
     throw new TypeError(`Attempted to send payload of invalid type '${typeof payload}'. Expected a string or Buffer.`)
   }
 
-  if (!reply.res.getHeader('Content-Length')) {
-    reply.res.setHeader('Content-Length', '' + Buffer.byteLength(payload))
+  if (!reply._headers['content-length']) {
+    reply._headers['content-length'] = '' + Buffer.byteLength(payload)
   }
 
   reply.sent = true
-  reply.res.end(payload)
+
+  res.writeHead(res.statusCode, reply._headers)
```

Comment:
You can avoid the ArgumentsAdaptorTrampoline in `res.writeHead` calls too by passing in a third argument.
https://github.com/nodejs/node/blob/master/lib/_http_server.js#L200

Reply:
Unfortunately if I pass `null` or `undefined` all tests fails.

Reply:
Oh that's too bad. I guess maybe the `writeHead` code is different for older versions of Node (it looks like it should work with the current code on master).

Reply:
Have you tried `res.writeHead(res.statusCode, null, reply._headers)`?
Or `res.writeHead(res.statusCode, http.STATUS_CODE[res.statusCode], reply._headers)`?

Reply:
@allevo can't make it to work, but I would like to land this before the 1st of March, and this can be worked out afterwards.

Reply:
There's another problem. Using the three parameters version on http2 a warning is emitted:
`(node:2103) UnsupportedWarning: Status message is not supported by HTTP/2 (RFC7540 8.1.2.4)`